### PR TITLE
feat: add pagination to sql reads

### PR DESCRIPTION
**Pull request by pablocampogo**
Add pagination to `GetTransactions` and `GetBlocks`, and implemented `GetTransactionsByAddress` also with pagination.

### CODE REVIEW

File: postgres-driver/postgres_driver.go [hunk 1]
```diff
@@ -127,12 +174,95 @@ func (d *PostgresDriver) WriteTransactions(txs []*indexer.Transaction) error {
...
-// TODO: add pagination
-func (d *PostgresDriver) ReadTransactions() ([]*indexer.Transaction, error) {
+// ReadTransactionsOptions optional parameters for ReadTransactions
+type ReadTransactionsOptions struct {
+	PerPage int
+	Page    int
+}
+
+// ReadTransactions returns transactions on the database with pagination
+// Optinal values defaults: page: 1, perPage: 1000
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:184]:
```suggestion
// Optional values defaults: page: 1, perPage: 1000
```

File: postgres-driver/postgres_driver.go [hunk 2]
```diff
@@ -127,12 +174,95 @@ func (d *PostgresDriver) WriteTransactions(txs []*indexer.Transaction) error {
...
+}
+
+// ReadTransactionsByFromAddressOptions optional parameters for ReadTransactionsByFromAddress
+type ReadTransactionsByFromAddressOptions struct {
+	PerPage int
+	Page    int
+}
+
+func validateAddress(fromAddress string) bool {
+	return addressRegex.MatchString(fromAddress) && len(fromAddress) == addressLength
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:231]: this sounds like an util function for pocket-go

File: postgres-driver/postgres_driver.go [hunk 3]
```diff
@@ -127,12 +174,95 @@ func (d *PostgresDriver) WriteTransactions(txs []*indexer.Transaction) error {
...
+}
+
+// ReadTransactionsByFromAddress returns transactions with given from address
+// Optinal values defaults: page: 1, perPage: 1000
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:235]:
```suggestion
// Optional values defaults: page: 1, perPage: 1000
```

File: postgres-driver/postgres_driver.go [hunk 4]
```diff
@@ -203,12 +333,40 @@ func (d *PostgresDriver) WriteBlock(block *indexer.Block) error {
 	return nil
 }
 
-// ReadBlocks returns all blocks on the database
-// TODO: add pagination
-func (d *PostgresDriver) ReadBlocks() ([]*indexer.Block, error) {
+// ReadBlocksOptions optional parameters for ReadBlocks
+type ReadBlocksOptions struct {
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:337]: why you have the same struct values with different names, I think you could use a single one called `GetPaginatedOptions` and use it whenever the only available body values are `perPage` and `page`

File: postgres-driver/postgres_driver.go [hunk 5]
```diff
@@ -203,12 +333,40 @@ func (d *PostgresDriver) WriteBlock(block *indexer.Block) error {
...
+	PerPage int
+	Page    int
+}
+
+// ReadBlocks returns all blocks on the database with pagination
+// Optinal values defaults: page: 1, perPage: 1000
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:343]:
```suggestion
// Optional values defaults: page: 1, perPage: 1000
```

File: postgres-driver/postgres_driver.go [hunk 6]
```diff
@@ -203,12 +333,40 @@ func (d *PostgresDriver) WriteBlock(block *indexer.Block) error {
...
+func (d *PostgresDriver) ReadBlocks(options *ReadBlocksOptions) ([]*indexer.Block, error) {
+	perPage := defaultPerPage
+	page := defaultPage
+
+	if options != nil {
```
Comment (rem1niscence) [inline postgres-driver/postgres_driver.go:348]: I've seen this body of checking options, and moving query several times, I think they could be extracted to a single function that takes as parameter the string query and does all this checks and fetching data, so could avoid this repetition unless you want to do something else besides that

**Review by rem1niscence — CHANGES_REQUESTED** [review]

File: postgres-driver/postgres_driver.go (hunk 4 above)
Comment (pablocampogo) [reply postgres-driver/postgres_driver.go:337]: It's so probable we'll need more optional values for each query, that's why I separated it

Comment (rem1niscence) [reply postgres-driver/postgres_driver.go:337]: all right

**Review by rem1niscence — APPROVED** [review]

File: postgres-driver/postgres_driver.go (hunk 2 above)
Comment (pablocampogo) [reply postgres-driver/postgres_driver.go:231]: U right

**Pull request merged**